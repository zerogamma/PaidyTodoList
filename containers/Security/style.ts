import { StyleSheet } from "react-native";
// Utils
import colors from "../../assets/colors/color";

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.primary,
    alignContent: "center",
    justifyContent: "center",
  },
  background: {
    position: "absolute",
    left: 0,
    right: 0,
    top: 0,
    height: "100%",
  },
  authWrapper: {},
});
