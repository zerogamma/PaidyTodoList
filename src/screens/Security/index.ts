export { Security as default } from "./Security";
