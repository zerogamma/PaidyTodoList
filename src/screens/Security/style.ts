import { StyleSheet } from "react-native";
// Utils
import colors from "../../../assets/colors/color";

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: colors.primary,
  },
  background: {
    position: "absolute",
    left: 0,
    right: 0,
    top: 0,
    height: "100%",
  },
  authWrapper: {
    bottom: 30,
    width: "100%",
    paddingHorizontal: 35,
    flexDirection: "column",
    justifyContent: "space-around",
    alignItems: "center",
  },
  description: {
    color: 'black',
    top: 200,
    fontSize: 30,
    position: "absolute"
  },
  buttonWrapper:{
    top: 450,
  },
  button: {
    width: 250,
    backgroundColor: colors.third,
    borderRadius: 60,
    paddingVertical: 15,
    paddingHorizontal: 15,
    color: colors.secondary,
    textAlign: 'center',
    fontSize: 25,
  },
});
