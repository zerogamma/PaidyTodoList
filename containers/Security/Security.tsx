import React, { useEffect, useState } from "react";
import * as LocalAuthentication from "expo-local-authentication";
import { View, Text, Button } from "react-native";
// components
import Main from "../../containers/Main";
import { LinearGradient } from "expo-linear-gradient";
import { styles } from "./style";
// Utils
import colors from "../../assets/colors/color";

export function Security() {
  const [biometrics, setBiometrcis] = useState<boolean>(false);
  const [renderContent, setRenderContent] = useState<boolean>(false);
  const [grantAccess, setGrantAccess] = useState<boolean>(false);

  useEffect(() => {
    (async () => {
      const compatible = await LocalAuthentication.hasHardwareAsync();
      setBiometrcis(compatible);
    })();
  }, []);

  const renderSecureContent = () => {
    (async () => {
      const auth = await LocalAuthentication.authenticateAsync();
      auth.success ? setGrantAccess(true) : setGrantAccess(false);
    })();
    setRenderContent(true);
  };

  return (
    <View style={styles.container}>
      <LinearGradient
        colors={colors.primaryGradient}
        style={styles.background}
      />
      {!biometrics ? (
        <Text>Face or Fingerprint scanner is not availbale on this device</Text>
      ) : renderContent && grantAccess ? (
        <Main />
      ) : (
        <View style={styles.authWrapper}>
          <Text>Set Authentication to proceed</Text>
          <Button title="Press to unlock" onPress={renderSecureContent} />
        </View>
      )}
    </View>
  );
}
