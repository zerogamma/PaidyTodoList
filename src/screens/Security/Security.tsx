import React, { useEffect, useState } from "react";
import * as LocalAuthentication from "expo-local-authentication";
import { View, Text, TouchableOpacity } from "react-native";
// components
import Main from "../Main";
import { LinearGradient } from "expo-linear-gradient";
import { styles } from "./style";
// Utils
import colors from "../../../assets/colors/color";
import ContextProvider from "../../context/ContextProvider";

export function Security() {
  const [biometrics, setBiometrcis] = useState<boolean>(false);
  const [renderContent, setRenderContent] = useState<boolean>(false);
  const [grantAccess, setGrantAccess] = useState<boolean>(false);
  const [facialRecognitionAvailable, setFacialRecognitionAvailable] =
    React.useState(false);
  const [fingerprintAvailable, setFingerprintAvailable] = React.useState(false);
  const [irisAvailable, setIrisAvailable] = React.useState(false);

  const checkSupportedAuthentication = async () => {
    const types = await LocalAuthentication.supportedAuthenticationTypesAsync();
    if (types && types.length) {
      const facial = types.includes(
        LocalAuthentication.AuthenticationType.FACIAL_RECOGNITION
      );
      const finger = types.includes(
        LocalAuthentication.AuthenticationType.FINGERPRINT
      );
      const eyes = types.includes(LocalAuthentication.AuthenticationType.IRIS);

      setFacialRecognitionAvailable(facial);
      setFingerprintAvailable(finger);
      setIrisAvailable(eyes);
      setBiometrcis(facial || finger || eyes);
    }
  };

  useEffect(() => {
    checkSupportedAuthentication();
  }, []);

  const getDescription = () => {
    if (facialRecognitionAvailable && fingerprintAvailable && irisAvailable) {
      return "Authenticate with Face ID, touch ID or iris ID";
    }
    if (facialRecognitionAvailable && fingerprintAvailable) {
      return "Authenticate with Face ID or touch ID";
    }
    if (facialRecognitionAvailable && irisAvailable) {
      return "Authenticate with Face ID or iris ID";
    }
    if (fingerprintAvailable && irisAvailable) {
      return "Authenticate with touch ID or iris ID";
    }
    if (facialRecognitionAvailable) {
      return "Authenticate with Face ID";
    }
    if (fingerprintAvailable) {
      return "Authenticate with touch ID ";
    }
    if (irisAvailable) {
      return "Authenticate with iris ID";
    }
  };

  const renderSecureContent = async () => {
    const auth = await LocalAuthentication.authenticateAsync();
    auth.success ? setGrantAccess(true) : setGrantAccess(false);
    setRenderContent(true);
  };

  return (
    <View style={styles.container}>
      <LinearGradient
        colors={colors.primaryGradient}
        style={styles.background}
      />
      {!biometrics ? (
        <Text>Face or Fingerprint scanner is not availbale on this device</Text>
      ) : renderContent && grantAccess ? (
        <ContextProvider>
          <Main />
        </ContextProvider>
      ) : (
        <View style={styles.authWrapper}>
          <Text style={styles.description}>{getDescription()}</Text>
          <TouchableOpacity style={styles.buttonWrapper} onPress={renderSecureContent} testID="addTask">
            <Text style={styles.button}>Press to unlock</Text>
          </TouchableOpacity>
        </View>
      )}
    </View>
  );
}
