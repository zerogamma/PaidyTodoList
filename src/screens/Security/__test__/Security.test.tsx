import React from "react";
import { Security } from "../Security";
import { render, fireEvent, cleanup } from "@testing-library/react-native";
import { act } from 'react-dom/test-utils';


const SecurityComponent = () => {
  return <Security />;
};

describe("<TaskComponent />", () => {
  afterEach(cleanup);

  it("renders correctly", () => {
    const container = render(<SecurityComponent />);
    expect(container).toMatchSnapshot();
  });
});
