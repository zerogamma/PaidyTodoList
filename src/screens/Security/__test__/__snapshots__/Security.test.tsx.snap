// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TaskComponent /> renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgba(92,162,255,1)",
      "flex": 1,
    }
  }
>
  <ViewManagerAdapter_ExpoLinearGradient
    proxiedProperties={
      Object {
        "colors": Array [
          4294967295,
          0,
        ],
        "endPoint": undefined,
        "locations": undefined,
        "startPoint": undefined,
      }
    }
    style={
      Object {
        "height": "100%",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  />
  <Text>
    Face or Fingerprint scanner is not availbale on this device
  </Text>
</View>
`;
